File: 03/03.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const TREE = '#'
const OPEN = '.'

type slope struct {
	right int
	down  int
}

func main() {
	const inputFile = "input.txt"
	input, _ := ioutil.ReadFile(inputFile)
	pattern := strings.Split(string(input), "\n")
	slopes := []slope{
		{right: 1, down: 1},
		{right: 3, down: 1},
		{right: 5, down: 1},
		{right: 7, down: 1},
		{right: 1, down: 2}}
	treeHits := []int{}
	for _, slope := range slopes {
		numberOfTreesHit := getNumberOfTreesHit(pattern, slope)
		fmt.Printf("Number of trees hit: %d (down %d, right %d)\n", numberOfTreesHit, slope.down, slope.right)
		treeHits = append(treeHits, numberOfTreesHit)
	}
	treeProduct := getProduct(treeHits)
	fmt.Printf("Product of trees hit: %d\n", treeProduct)
}

func getNumberOfTreesHit(pattern []string, slope slope) int {
	row := 0
	column := 0
	treesHit := 0
	for row < len(pattern) {
		if hitTree(row, column, pattern) {
			treesHit++
		}
		row += slope.down
		column += slope.right
	}
	return treesHit
}

func hitTree(row int, column int, pattern []string) bool {
	moduloColumn := column % len(pattern[0])
	if pattern[row][moduloColumn] == TREE {
		return true
	}
	return false
}

func getProduct(ns []int) int {
	nProduct := ns[0]
	for _, n := range ns[1:] {
		nProduct = nProduct * n
	}
	return nProduct
}

File: 03/03.hs

tree = '#'

data Slope = Slope { right::Int, down::Int } deriving (Show)

main :: IO ()
main = do
  inputs <- lines <$> readFile "input.txt"
  let rowLen = length (head inputs)
  let depth = length inputs
  let treeIndices = getTreeIndices inputs
  let slopes = [Slope 3 1, Slope 1 1, Slope 5 1, Slope 7 1, Slope 1 2]
  let steps = generateSteps (depth-1) (head slopes)
  let treesHit = length (filter (uncurry (hitTree treeIndices rowLen)) steps)
  putStrLn ("Part 1 trees hit: " ++ show treesHit)

  let stepss = map (generateSteps (depth-1)) slopes
  let treesHitPerSlope = map (length . filter (uncurry (hitTree treeIndices rowLen))) stepss
  let treesHitProduct = product treesHitPerSlope
  putStrLn ("Part 2 product of trees hit: " ++ show treesHitProduct)


generateSteps :: Int -> Slope -> [(Int, Int)]
generateSteps depth slope = do
  let rows = [0, (down slope) .. depth]
  let columns = [0,(right slope)..depth * right slope]
  zip rows columns

getTreeIndices :: [String] -> [[Int]]
getTreeIndices pattern = do
  let patternWithIndices = map (\line -> zip line [0..(length line)]) pattern
  let treesWithIndices = map (filter (\(c,_) -> c == tree)) patternWithIndices
  let treeIndices = map (map snd) treesWithIndices
  treeIndices

hitTree :: [[Int]] -> Int -> Int -> Int -> Bool
hitTree treeIndices rowLen row col = do
  let moduloCol = col `mod` rowLen
  moduloCol `elem` (treeIndices!!row)
